brave/brave-core
# Refactor unblinded tokens

Resolves https://github.com/brave/brave-browser/issues/16005

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

- Confirm tokens are refilled after successfully fetching issuers from `GET /v1/issuers`
- Confirm tokens are refilled after successfully redeeming a token
- Confirm tokens are refilled after failing to redeem a token redeeming a token
- Confirm users are rewarded for viewing ads
- Confirm tokens are cashed-out
- Confirm `creativeType` (i.e. `ad_notification`, `new_tab_page_ad`, `inline_content_ad` or `promoted_content_ad`) is not included in the payload when tokens are cashed-out
- Confirm issuers are rejected if any payment issuer has more than two public keys with the same value
- Confirm issuers are rejected if the payment issuer is missing from  `GET /v1/issuers/` endpoint
- Confirm issuers are rejected if issuers have not changed since the last update
- Confirm issuers are fetched from `GET /v1/issuers/` endpoint on browser launch
- Confirm issuers are fetched when ads are enabled
- Confirm issuers are not fetched when updating the catalog
- Confirm issuers are fetched every x milliseconds (see `ping` in `GET /v1/issuers/` response)
- Confirm "like" button (thumbs up) on 30-day ads history sends an upvote confirmation to the server
- Confirm "dislike" button (thumbs down) on 30-day ads history sends a downvote confirmation to the server
- Confirm "save" button on 30-day ads history sends a saved confirmation to the server
- Confirm "mark as inappropriate" button on 30-day ads history sends a flag confirmation to the server
- Confirm clicking an ad sends a click confirmation to the server
- Confirm converting an ad sends a conversion confirmation to the server
- Confirm landing an ad (transferred) sends landed a confirmation to the server
- Confirm `unblinded_payment_tokens` node in `confirmations.json` is migrated to include `ad_type` (empty string), `confirmation_type` (empty string) and `transaction_id` (random v4 UUID)
- Confirm viewed ads are shown within 30-day ads history
- Confirm issuers are persisted in `confirmations.json`
- Confirm `unblinded_payment_tokens` node in `confirmations.json` has `ad_type`, `confirmation_type` and `transaction_id` for new entries
- Confirm catalog is downloaded from `GET /v9/catalog` endpoint

NOTE: `Failed to get issuers` will be shown after upgrading because the `issuers` node will only be created after successfully downloading issuers from `GET /v1/issuers/` for the first time. See the spec for more detail
